FuelLabs/fuel-indexer
# enhancement: dockerize examples + update plugin

### Description
- This PR does a few things 
  - Adds docker functionality to examples
  -  Updates each examples README file with proper usage steps 
  -  Updates the plugin to support this dockerization
  - Adds a dockerized version of our test `fuel-node`
  - Renames `hello-bin` to `hello-world-data` for clarity
    - Plugin commands `build`, `deploy`, `remove` now support `--path` and (new) `--manifest` options
      - `--path` is the path to the indexer project (else the current directory is used)
        - MImics forc-deploy's [`--path`](https://github.com/FuelLabs/sway/blob/master/forc-plugins/forc-client/src/ops/deploy/cmd.rs#L9)
      -  `--manifest` is the name of the manifest file - else it defaults to the convention `{project_name}.manifest.yaml`
      -  The new `--output-dir-root` flag (flexible on the naming) is used in niche cases where you want to output your build artifact(s) and deploy those artifacts from a directory that is _not_ the indexer project directory
          - This allows (for example) us to build and deploy examples from `fuel-indexer` or `fuel-indexer/examples` or `fuel-indexer/examples/$example_project_name` 
            - Again, I'm open to ideas slash flexible on this
  - Adds `stop_idle_indexers` config option
    - Indexers by default run forever -- that's good for production, but not good for CI
      - If passed, this flag will stop a running indexer after `N` amount of empty block requests -- where `N` is  a constant (but can be configured later if need be)
 
### Testing steps
- For both examples, follow the instructions in the README
- For the hello-world example, please see the caveat below
  
- [ ] Testing the block explorer example
  - Just follow the README exactly 
- [ ] Testing the hello-world example
  - In the docker-compose swap the `image` in the `fuel-indexer` service for a local `fuel-indexer/local:latest` image
  - If you don't have this image locally, build it first with `docker build -t fuel-indexer/local:latest -f deployment/Dockerfile . --no-cache` 
    -  The image in the Github registry does not have these PR changes
- [ ] There's testing for the plugin functionality as well
  -  Try `build` `deploy` `remove` from an indexer project directory with no `--path` or `--manifest`
  -  Try `build` `deploy` `remove` from _outside_ the indexer project directory with a `--path`
  -   Rename the default indexer manifest to `foo.yaml` and try to `build` `deploy` and `remove` using `--manifest foo.yaml`
